added seebeck generator

niklaswimmer opened the pull request:
whoop
I guess everything should work. I hope that I haven't forgot a feature.
Textures are temporary (I don't like that black/purple look)

niklaswimmer committed:
added seebeck generator

niklaswimmer committed:
small changes to the power production
removed few and added more comments

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/Blocks.kt:
```diff
@@ -18,6 +18,10 @@ internal object Blocks {
         MadBlock().setCreativeTab(mainItemGroup).setTranslationKey("ematter.molecular_assembler_device").setHardness(8.0F).apply { this.setHarvestLevel("pickaxe", 2) }
     }
+    //I guess that's how one is supposed to do it? At least it works :)
+    val seebeckGenerator = register( "seebeck_generator" ) {
```
No spaces between parentheses and arguments please

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/feature/sbg/SbgBlock.kt:
```diff
@@ -0,0 +1,31 @@
+     * Gets called whenever a neighbor block changes and calls the recalculateNeighbors() fun inside SbgTileEntity
+     */
+    override fun neighborChanged( state: IBlockState , worldIn: World , pos: BlockPos , blockIn: Block , fromPos: BlockPos ) {
+        if( !worldIn.isRemote ) {
```
Prefer early escape rather than indenting: it makes the code way cleaner from a reader's perspective

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/feature/sbg/SbgTileEntity.kt:
```diff
@@ -0,0 +1,200 @@
+     * and the [threshold] is at 0.
+     */
+    override fun update() {
+        if( !super.world.isRemote ) {
```
Why are you invoking `super`? There is no need. You can just use `this`

niklaswimmer replied:
Is `super` slower or something like that? I thought invoking `super` is better for readability but I'll change it

TheSilkMiner replied:
It's not slower, actually `this.world == super.world`. It simply makes more sense when reading: "if this world is not remote" rather than "if the super world is not remote". What's a `super` world?
I mean, it doesn't change the meaning that much, but... yeah

niklaswimmer replied:
ohh ok. I'll think about that more often now

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/feature/sbg/SbgTileEntity.kt:
```diff
@@ -0,0 +1,200 @@
+        if( !super.world.isRemote ) {
+            if( this.tempDifference != 0u ) {
+                if( this.threshold > 0u ) {
+                    this.threshold--
```
`--this.treshold`

niklaswimmer replied:
You mean inside the if or `if( --this.threshold > 0u )`

TheSilkMiner replied:
```kotlin
if (this.threshold > 0U) --this.treshold
```

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/feature/sbg/SbgTileEntity.kt:
```diff
@@ -0,0 +1,200 @@
+     * Ampere gets generated every [BURST_TIME] but in an amount that equals generation every tick.
+     */
+    private fun generatePower() {
+        val nextPowerProduced: Double = this.tempDifference.toDouble() * 0.005 * this.effectiveness * POWER_MULTIPLIER
```
No need to specify the type directly, this can be automatically inferred

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/feature/sbg/SbgTileEntity.kt:
```diff
@@ -0,0 +1,200 @@
+
+        for( direction in Direction.values() ) {
+            //the top is only for energy so don't have to be checked here
+            if( direction != Direction.UP ) {
```
Can't this be excluded from the loop directly?

niklaswimmer replied:
I guess. Didn't know how but I'll look into it.

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/feature/sbg/SbgTileEntity.kt:
```diff
@@ -0,0 +1,200 @@
+     */
+    private fun getNeighborAtDir( direction: Direction ): Block {
+        return when( direction ) {
+            Direction.NORTH -> super.pos.north()
```
No need to `super`

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/feature/sbg/SbgTileEntity.kt:
```diff
@@ -0,0 +1,200 @@
+     * @param direction the direction of the neighbor
+     */
+    private fun getNeighborAtDir( direction: Direction ): Block {
+        return when( direction ) {
```
You are already returning: consider using an inlined function for this, rather than one with an actual body

niklaswimmer replied:
What exactly do you mean with this?

TheSilkMiner replied:
```kotlin
private fun getNeighborAtDir(direction: Direction) = when (direction) { ..... }
```
No need for it to be inline, but it looks cleaner IMO

niklaswimmer replied:
ahh you mean like that. Yes I can do this

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/shared/TemperatureSources.kt:
```diff
@@ -0,0 +1,25 @@
+package net.thesilkminer.mc.ematter.common.shared
```
This should be in the Seebeck feature group, but it really doesn't matter
This will get moved eventually

TheSilkMiner commented on src/main/resources/assets/ematter/blockstates/seebeck_generator.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "variants": {
+    "normal": {
```
`"normal"` should be `""` to respect the 1.13 blockstate format

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/shared/TemperatureSources.kt:
```diff
@@ -0,0 +1,25 @@
+    LAVA_FLOW( Blocks.FLOWING_LAVA , 1000 ), //same as for the water
+    MAGMA( Blocks.MAGMA , 1400 )
+
+}
```
Missing newline at end of file

TheSilkMiner commented on src/main/resources/assets/ematter/models/block/seebeck_generator.json:
```diff
@@ -0,0 +1,12 @@
+    "south": "ematter:blocks/seebeck_generator/side",
+    "west": "ematter:blocks/seebeck_generator/side"
+  }
+}
```
Missing newline at end of file (though yet again this is temporary, it'll be replaced with a BlockBench alternative so...)

TheSilkMiner commented on src/main/resources/assets/ematter/models/item/seebeck_generator.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "parent": "ematter:block/seebeck_generator"
+}
```
Missing newline at end of file

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/feature/sbg/SbgTileEntity.kt:
```diff
@@ -0,0 +1,200 @@
+        private const val MAX_CAPACITY: ULong = 3_141uL
+
+        //next one is here in case you want to make this changeable through a config file in the future
+        private const val POWER_MULTIPLIER: Double = 1.0
```
This should be placed in a configuration file, which isn't set up yet
Consider this on hold

niklaswimmer replied:
So, should I just remove the val or leave it for now?

TheSilkMiner replied:
It's fine for now: this is more of a reminder for the future

niklaswimmer replied:
ok. I'll mark that with a comment on that line

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/feature/sbg/SbgTileEntity.kt:
```diff
@@ -0,0 +1,200 @@
+            if( direction != Direction.UP ) {
+                val neighbor: Block = this.getNeighborAtDir( direction )
+                for( temperatureSource in TemperatureSources.values() ) {
+                    if( neighbor == temperatureSource.block ) {
```
You could use `find` for this

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/feature/sbg/SbgTileEntity.kt:
```diff
@@ -0,0 +1,200 @@
+            this.burstTimer = BURST_TIME
+            this.powerProduced = 0u
+        }
+        else {
```
`else` should be on the same line as the ending `if` brace

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/feature/sbg/SbgTileEntity.kt:
```diff
@@ -0,0 +1,200 @@
+            //for every heat source that has no coolant the effectiveness gets reduced (one coolant for two heat sources)
+            if( heatSources > coolants * 2 ) {
+                for( i in 1..( heatSources - coolants * 2 ) ) {
+                    this.effectiveness -= 0.2
```
I feel this doesn't need to be a loop

niklaswimmer replied:
I'll rethink the logic I used there.

TheSilkMiner commented on src/main/kotlin/net/thesilkminer/mc/ematter/common/feature/sbg/SbgTileEntity.kt:
```diff
@@ -0,0 +1,200 @@
+                        this.generatePower()
+                    }
+                }
+            }
```
You aren't pushing the produced energy anywhere. Is this intended for a later addition when cables are actually a thing or was it an oversight?

niklaswimmer replied:
Actually I haven't thought about that at all. But I guess this can be added later on when cables or such are a thing?
I'll look into that and try to figure out how to push it somewhere and if I can't I'll ask you again later.

TheSilkMiner reviewed (changes requested):
After all it's a very well done job, and thank you for the contribution. Just some minor nitpicks here and there related to code style.
The logic for the Seebeck Generator seems spot on and... well... apart from a few nitpicks here and there, which may be due to a small difference in the approach used, that's fine.
Again, good job

niklaswimmer committed:
requested changes
sbg still works the same only changes to how it is coded and general code style

niklaswimmer committed:
sbg can now push power out on the top
